### PR TITLE
Windows support

### PR DESCRIPTION
**Pull request by albertvaka**
### What does this PR do?
 
- Adds a class to download and install the agent 6 from an MSI, similar to the ones we have for the APT and RPM packages. (Note: this adds a dependency on powershell).
- The main `datadog_agent` class now uses the MSI install method on Windows, while the code that creates the config files is shared between Windows and Linux.
- The integration-specific classes that create integrations config files all compile on Windows, but haven't been tested.
- The `datadog_agent::reports` class hasn't been ported. Trying to use it on Windows throws an error.
- Paths, usernames, file permissions... have been extracted into variables, so they can have different values on Windows and on Linux. These variables are defined in `params.pp` for puppet to use, and in `spec_helper.rb` for rspec.
- Makes test work on Windows and adds one Travis job that tests on Windows. (Note: it takes ~20 min vs ~5 min on Linux, because there is no official Travis image for Ruby on Windows and we need to install a bunch of packages).
- Simplifies the tests by relying less on local variables and facts, and more on `default_facts` and global variables from `spec_helper.rb`.

### Motivation

- A few users and customers have expressed interest in using Puppet on Windows.

### Additional notes

- This supports both installing and uninstalling the agent. Since on Linux we don't support uninstalling, this is specified with the parameter `win_ensure` instead of just `ensure`. (Alternative: call it `ensure` and error out in Linux?).
- Installing Agent 5 is not supported and throws an error.

Rebased from https://github.com/DataDog/puppet-datadog-agent/pull/552

### CODE REVIEW

File: metadata.json [hunk 1]
```diff
@@ -32,6 +32,10 @@
     {
       "operatingsystem": "Ubuntu",
       "operatingsystemrelease": ["14.04", "16.04", "17.04", "17.10", "18.04", "18.10"]
+    },
+    {
+      "operatingsystem": "Windows",
```
Comment (albertvaka) [inline metadata.json]: We might want to leave this out for the first release, so we don't announce it as compatible with Windows yet. Comments?

File: .travis.yml [hunk 2]
```diff
@@ -36,3 +38,14 @@ matrix:
...
     - rvm: 2.2.3
       env: PUPPET_VERSION="~> 6.5.0"
+  include:
+      os: windows
+      language: bash # There is no support for ruby on windows yet (2019/09)
+      install:
+      - choco install ruby
+      - choco install msys2
+      - ridk.cmd exec pacman -S --noconfirm --needed base-devel mingw-w64-x86_64-toolchain
+      - powershell kill -n gpg-agent # Started by msys, if we don't kill it the travis job stays running forever (found by running 'ps -W')
```
Comment (truthbk) [inline .travis.yml]: 👍

File: manifests/init.pp [hunk 3]
```diff
@@ -706,16 +733,35 @@
...
+
+    if ($::operatingsystem == 'Windows') {
+
+      file { 'C:/ProgramData/Datadog':
+        ensure   => directory
+      }
+
+      file { 'C:/ProgramData/Datadog/datadog.yaml':
+        owner   => $dd_user,
+        group   => 'S-1-5-32-544', #Administrators
```
Comment (truthbk) [inline manifests/init.pp]: Looks good to me, but if you haven't, can you run this by @derekwbrown?

File: manifests/install_integration.pp [hunk 4]
```diff
@@ -6,19 +6,25 @@
 
   include datadog_agent
 
+  if $::operatingsystem == 'Windows' {
+    $agent_binary = 'C:/Program Files/Datadog/Datadog Agent/embedded/agent.exe'
```
Comment (truthbk) [inline manifests/install_integration.pp]: should perhaps these be in the `params.pp` file?

File: manifests/reports.pp [hunk 5]
```diff
@@ -22,35 +22,43 @@
...
-  $rubydev_package = $datadog_agent::params::rubydev_package
-
-  # check to make sure that you're not installing rubydev somewhere else
-  if ! defined(Package[$rubydev_package]) {
-    package {$rubydev_package:
-      ensure => installed,
-      before => Package['dogapi']
+  if ($::operatingsystem == 'Windows') {
+
+    fail('Reports are not supported on Windows')
```
Comment (truthbk) [inline manifests/reports.pp]: I think this message is a tad bit misleading - windows puppet clients will generate their reports, send them to the `master`, and the `master` node should report them. I guess what we're trying to say here something like _"Reporting is not supported from a windows host"_ (or something along those lines)?

File: manifests/windows/agent6.pp [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  String $service_ensure = 'running',
+  String $baseurl = $datadog_agent::params::agent6_default_repo,
+  String $msi_location = 'c:/tmp',
+  String $api_key = $datadog_agent::api_key,
+  String $hostname = $datadog_agent::host,
+  String $service_name = $datadog_agent::service_name_win,
+  Array  $tags = $datadog_agent::tags,
+  Boolean $service_enable = true,
+  Enum['present', 'absent'] $ensure = 'present',
+) inherits datadog_agent::params {
```
Comment (truthbk) [inline manifests/windows/agent6.pp]: 🍰 for the typed params

File: metadata.json [hunk 7]
```diff
@@ -68,6 +72,10 @@
     {
       "name": "puppetlabs/yumrepo_core",
       "version_requirement": ">=1.0.3 < 2.0.0"
+    },
+    {
+      "name": "puppetlabs/powershell",
+      "version_requirement": ">=2.3.0 <3.0.0"
```
Comment (truthbk) [inline metadata.json]: Technically `puppet-datadog-agent` supports some older puppets like `4.6.2`, I believe powershell `>=2.3.0` only officially supports `>=4.7.x`. That said, I don't think this is a problem, it's a new feature, and those older puppets are EOL. It should be fine, if anything breaks with older puppets on windows, to affirm we do *NOT* support `<4.7.0` on win?

File: metadata.json (hunk 1 above)
Comment (truthbk) [reply metadata.json]: Makes sense, or release it as a beta feature? It shouldn't affect linux otherwise.

File: spec/spec_helper.rb [hunk 8]
```diff
@@ -5,26 +5,69 @@
 
 DEBIAN_OS = %w(Ubuntu Debian)
 REDHAT_OS = %w(RedHat CentOS Fedora Amazon Scientific)
-ALL_OS = DEBIAN_OS + REDHAT_OS
+WINDOWS_OS = %w(Windows)
+
+if RSpec::Support::OS.windows?
```
Comment (truthbk) [inline spec/spec_helper.rb]: Fantastic stuff with the helper! So much cleaner all across the spec tests. Big thanks for this 🙇 

I guess we didn't find a way to use the params here. It's a shame but no worries.

**Pull request merged**